File: skald/agents/chat_agent/preprocessing.py
from typing import Any, Dict, List, Optional

from skald import settings
from skald.embeddings.vector_search import memo_chunk_vector_search
from skald.models.project import Project
from skald.services.embedding_service import EmbeddingService
from skald.services.rerank_service import RerankService
from skald.utils.filter_utils import MemoFilter


def _chunk_vector_search(
    query: str, project: Project, filters: Optional[List[MemoFilter]] = None
) -> List[Dict[str, Any]]:
    from skald.models.memo import MemoSummary

    embedding_vector = EmbeddingService.generate_embedding(query, usage="search")
    chunk_results = memo_chunk_vector_search(
        project, embedding_vector, settings.VECTOR_SEARCH_TOP_K, filters=filters
    )

    rerank_data = []
    for chunk_result in chunk_results:
        chunk = chunk_result["chunk"]
        memo = chunk.memo

        # Try to get summary, use fallback if it doesn't exist
        try:
            summary = memo.summary
        except MemoSummary.DoesNotExist:
            summary = "[Summary not yet generated]"

        rerank_snippet = f"Title: {memo.title}\n\nFull content summary: {summary}\n\nChunk content: {chunk.chunk_content}\n\n"
        rerank_data.append(rerank_snippet)

    # split into batches of 25 to ensure we're under the 32k token limit
    rerank_data_batches = [
        rerank_data[i : i + 25] for i in range(0, len(rerank_data), 25)
    ]

    results = []
    for batch in rerank_data_batches:
        rerank_result = RerankService.rerank(query, batch)
        results.extend(rerank_result)

    return results


def prepare_context_for_chat_agent(
    query: str, project: Project, filters: Optional[List[MemoFilter]] = None
) -> List[Dict[str, Any]]:
    """
    Async version of prepare_context_for_chat_agent that processes rerank batches in parallel.

    Args:
        query: The search query
        project: The project to scope the search to

    Returns:
        List of reranked results
    """

    results = _chunk_vector_search(query, project, filters)

    # sort all results by relevance score
    results.sort(key=lambda x: x.relevance_score, reverse=True)

    return results[: settings.POST_RERANK_TOP_K]

File: express-server/src/agents/chatAgent/preprocessing.ts
import { Project } from '../../entities/Project'
import { MemoFilter } from '../../lib/filterUtils'
import { EmbeddingService } from '../../services/embeddingService'
import { RerankService } from '../../services/rerankService'
import { memoChunkVectorSearch } from '../../embeddings/vectorSearch'
import { VECTOR_SEARCH_TOP_K, POST_RERANK_TOP_K } from '../../settings'
import { DI } from '../../index'

interface RerankResult {
    index: number
    document: string
    relevance_score: number
}

async function chunkVectorSearch(query: string, project: Project, filters?: MemoFilter[]): Promise<string[]> {
    // Generate embedding for the query
    const embeddingVector = await EmbeddingService.generateEmbedding(query, 'search')

    // Perform vector search
    const chunkResults = await memoChunkVectorSearch(project, embeddingVector, VECTOR_SEARCH_TOP_K, 0.55, filters)

    // Build rerank data
    const rerankData: string[] = []
    for (const chunkResult of chunkResults) {
        const chunk = chunkResult.chunk
        const memoId = chunk.memo

        // Load the full memo
        const memo = await DI.em.getRepository('Memo').findOne({ uuid: memoId as any })

        if (!memo) {
            continue
        }

        // Try to get summary, use fallback if it doesn't exist
        let summary = '[Summary not yet generated]'
        try {
            const memoSummary = await DI.em.getRepository('MemoSummary').findOne({ memo: memoId as any })
            if (memoSummary) {
                summary = (memoSummary as any).summary
            }
        } catch (error) {
            // Summary doesn't exist, use default
        }

        const rerankSnippet = `Title: ${(memo as any).title}\n\nFull content summary: ${summary}\n\nChunk content: ${chunk.chunk_content}\n\n`
        rerankData.push(rerankSnippet)
    }

    // Split into batches of 25 to ensure we're under the 32k token limit
    const rerankDataBatches: string[][] = []
    for (let i = 0; i < rerankData.length; i += 25) {
        rerankDataBatches.push(rerankData.slice(i, i + 25))
    }

    // Rerank all batches
    const results: RerankResult[] = []
    for (const batch of rerankDataBatches) {
        const rerankResult = await RerankService.rerank(query, batch)
        results.push(...rerankResult)
    }

    return results.map((r) => r.document)
}

export async function prepareContextForChatAgent(
    query: string,
    project: Project,
    filters?: MemoFilter[]
): Promise<RerankResult[]> {
    const results = await chunkVectorSearch(query, project, filters)

    // Convert strings back to rerank results with scores
    const rerankResults: RerankResult[] = results.map((doc, index) => ({
        index,
        document: doc,
        relevance_score: 1.0 - index * 0.01, // Simple decreasing score
    }))

    // Sort by relevance score
    rerankResults.sort((a, b) => b.relevance_score - a.relevance_score)

    // Return top K
    return rerankResults.slice(0, POST_RERANK_TOP_K)
}
